### ISSUE TITLE
Uncaught TypeError: Cannot call a class as a function @ v0.5.2 / v0.5.3

### ISSUE THREAD

**Issue by whitetrefoil** [HerringtonDarkholme/av-ts#17]
_When with v0.5.1 it's all good._

The error was caused here:
```js
_classCallCheck(this, HelloComponent); // then

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {  // false here
    throw new TypeError("Cannot call a class as a function");
  }
}
```

I compared the built code and the only difference is:

```js
// v0.5.1 (working)

function decorate(cls) {
  Component.inDefinition = true;
  // workaround flow inference
  var instance = undefined;
  try {
    instance = new cls();
  }
  finally {
    Component.inDefinition = false;
  }
```
```js
// v0.5.2 (broken)

function decorate(cls) {
  Component.inDefinition = true;
  var instance = Object.create(Vue.prototype);
  Object.defineProperty(instance, '_init', {
    value: util_1.NOOP, enumerable: false
  });
  try {
    cls.call(instance); // this calls _classCallCheck
  }
  finally {
    Component.inDefinition = false;
  }
```

I'm using vue@2.0.5, vue-loader@9.9.1
I'm **not** using vue-ts-loader, but awesome-typescript-loader and separated template like below:
```html
<style lang="sass" src="./app.component.sass" />
<template lang="pug" src="./app.component.pug" />
<script lang="ts" src="./app.component.ts" />
```
```ts
import { Vue, Component } from 'av-ts'

@Component({
  name: 'app',
})
export default class AppComponent extends Vue {
}
```

I've put my code here: https://github.com/whitetrefoil/vue-ts-arch/tree/av-ts-052

**Comment by HerringtonDarkholme** [comment]
Are you using babel? The error comes from babel transpilation. The `cls.call` is important for supporting Mixin and component extends.

The best way is skip babel, just use ts to transpile.

**Comment by HerringtonDarkholme** [comment]
Can you manually try this?

```ts

function decorate(cls) {
  Component.inDefinition = true;
  var instance = Object.create(cls.prototype);
  Object.defineProperty(instance, '_init', {
    value: util_1.NOOP, enumerable: false
  });
  try {
    cls.call(instance); // this calls _classCallCheck
  }
  finally {
    Component.inDefinition = false;
  }
```

**Comment by HerringtonDarkholme** [comment]
Fixed in 0.5.4

**Comment by whitetrefoil** [comment]
So quick... I'm still typing... Σ( ° △ °|||)︴
Yes, I am using babel to process `async/await => bluebird.coroutine`.
I'll try the new one, thanks!

**Comment by whitetrefoil** [comment]
Confirmed fix in 0.5.4!

**Comment by HerringtonDarkholme** [comment]
This is a known issue in vue-class-component https://github.com/vuejs/vue-class-component/pull/32.
I just forget it when implementing new class instance creation (´・ω・｀), so sorry.

**Comment by HerringtonDarkholme** [comment]
@whitetrefoil I find you are building an full blown app in Vue + TS. If you don't mind, I keenly invite you to take [kilimanjaro](https://github.com/HerringtonDarkholme/kilimanjaro), a typed vuex fork. I had a hard time when type checking vuex.

I would not recommend any one to use kilimanjaro in prodcution yet(except myself). But I'm curious about other users' opinion and feeling about it.

**Comment by whitetrefoil** [comment]
@HerringtonDarkholme I've already tried kilimanjaro in a small scope refactoring of a existing small project a few days ago ;)

I'm still considering to use kilimanjaro, but since I've just learnt TS for no longer than 1 week and I'm not yet very familiar with Vue(x), a full stack of libraries is too difficult to me.

I'll pick up them one by one, not too hurry. wwwww

**Comment by IsaacSee** [comment]
With v0.5.5 I still have `TypeError: Class constructor CLASSNAME cannot be invoked without 'new'`.
I am using `ts-vue-loader` and `babel`.

**Comment by HerringtonDarkholme** [comment]
Please provide valid reproduction code.

**Comment by IsaacSee** [comment]
Thanks for your quick answer.
It happends inside a quite large project, I am trying to migrate from another ts-component lib to `av-ts`.
I'll try to find time to extract a reproduction snippet.

**Comment by IsaacSee** [comment]
Just realized that I was not using babel !
Using the `babel-loader` after the `vue-ts-loader` did it.
` {
                test: /\.ts$/,
                loader: 'babel!vue-ts'
            },`

**Comment by IsaacSee** [comment]
I imagine this does not make much sense on a reproduction purpose but I am not yet familiar with webpack/typescript.

**Comment by whitetrefoil** [comment]
Oops... I met the very same problem now... The same code just worked yesterday and I cannot find out the difference... Assuming some dependency upgrades caused it.

```
Uncaught TypeError: Class constructor SiteFooter cannot be invoked without 'new'
    at decorate (index.js:9687)
    at __decorate (index.js:9497)
    at Object.<anonymous> (index.js:9506)
    at __webpack_require__ (polyfills.js:584)
    at fn (polyfills.js:108)
    at Object.<anonymous> (index.js:9160)
    at __webpack_require__ (polyfills.js:584)
    at fn (polyfills.js:108)
    at Object.<anonymous> (index.js:8013)
    at __webpack_require__ (polyfills.js:584)
    at fn (polyfills.js:108)
    at Object.<anonymous> (index.js:7658)
    at __webpack_require__ (polyfills.js:584)
    at fn (polyfills.js:108)
    at Object.<anonymous> (index.js:7)
    at __webpack_require__ (polyfills.js:584)
```

```js
	function Component_(meta) {
	    if (meta === void 0) { meta = {}; }
	    function decorate(cls) {
	        Component.inDefinition = true;
	        var instance = Object.create(cls.prototype);
	        Object.defineProperty(instance, '_init', {
	            value: util_1.NOOP, enumerable: false
	        });
	        try {
	            cls.call(instance);  // <= TypeError here
	        }
	        finally {
	            Component.inDefinition = false;
	        }
	        var proto = cls.prototype;
	        var options = makeOptionsFromMeta(meta, cls['name']);
	        var _a = getKeys(proto), internalKeys = _a.internalKeys, normalKeys = _a.normalKeys;
	        for (var _i = 0, internalKeys_1 = internalKeys; _i < internalKeys_1.length; _i++) {
	            var protoKey = internalKeys_1[_i];
	            collectInternalProp(protoKey, proto, instance, options);
	        }
	        for (var _b = 0, normalKeys_1 = normalKeys; _b < normalKeys_1.length; _b++) {
	            var protoKey = normalKeys_1[_b];
	            collectMethodsAndComputed(protoKey, proto, options);
	        }
	        // everything on instance is packed into data
	        collectData(cls, Object.keys(instance), options);
	        var Super = findSuper(proto);
	        return Super.extend(options);
	    }
	    return decorate;
	}
```

```js
/* 131 */
/***/ function(module, exports, __webpack_require__) {

	"use strict";
	var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
	    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
	    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
	    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	    return c > 3 && r && Object.defineProperty(target, key, r), r;
	};
	var __metadata = (this && this.__metadata) || function (k, v) {
	    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
	};
	const av_ts_1 = __webpack_require__(132);
	let SiteFooter = class SiteFooter extends av_ts_1.Vue {
	};
	SiteFooter = __decorate([
	    av_ts_1.Component({
	        name: 'site-footer',
	    }),
	    __metadata("design:paramtypes", [])
	], SiteFooter);
	Object.defineProperty(exports, "__esModule", { value: true });
	exports.default = SiteFooter;
```

Meanwhile, I also met a wired UglifyJs error during build (also happens only today):
```
Error: index-5b511b88878979cc800c.js from UglifyJs
SyntaxError: Unexpected token: name (SiteFooter) [./~/awesome-typescript-loader/dist/entry.js?tsconfig=tsconfig.json!./src/components/site-footer.ts:12,0]chunks/1-35504d298cbc131b0d92.chunk.js from UglifyJs
```
Not sure if they have any relationship...

The building settings of my project are almost the same as [https://github.com/whitetrefoil/vue-ts-arch/](), but this base project just work while my actual project doesn't...

I'm still digging & will report once I find any clue about it...

**Comment by HerringtonDarkholme** [comment]
@whitetrefoil Thanks for reporting! Some thoughts I have come up, you can try different babel version and av-ts version. I do have an update in av-ts but I don't think it is related to your problem.

**Comment by whitetrefoil** [comment]
@HerringtonDarkholme I'm trying to recover my previous dependency list from my laptop which I haven't used since last success build.  I'm also suspecting webpack / yarn / ts...

**Comment by whitetrefoil** [comment]
@HerringtonDarkholme Finally, after I deleted all `awesome-typescript-loader` cache files (`.awcache`, etc.), everything back to normal...

I'm afraid this may not help @IsaacSee 's issue since his using `vue-ts-loader`...

**Comment by whitetrefoil** [comment]
@IsaacSee I saw babel also has its own cache, maybe worthy to do a cleanup.

**Comment by HerringtonDarkholme** [comment]
@whitetrefoil I'll hack av-ts until it conforms ES6 semantics.

https://tc39.github.io/ecma262/#sec-ecmascript-function-objects-call-thisargument-argumentslist

**Comment by HerringtonDarkholme** [comment]
Fixed again in 0.5.6. Now the code is ES6 compatible.

**Comment by whitetrefoil** [comment]
Doesn't print any red line in my project's console ~
